node/core/pvf/src/artifacts.rs

```diff
@@ -0,0 +1,288 @@
+
+		let file_name = file_name.strip_prefix(Self::PREFIX)?;
+		let code_hash = Hash::from_str(file_name).ok()?;
+
+		Some(Self { code_hash })
+	}
+
+	/// Returns the expected path to this artifact given the root of the cache.
+	pub fn path(&self, cache_path: &Path) -> PathBuf {
+		let file_name = format!("{}{}", Self::PREFIX, self.code_hash.to_string());
+		cache_path.join(file_name)
+	}
+}
+
+#[test]
+fn ensure_wasmtime_version() {
+	assert_eq!(
+		wasmtime_jit::VERSION,
+		"0.22.0",
+		"wasmtime version is updated. Check the prefix.",
+	);
+	// If the version bump is significant, change `ArtifactId::PREFIX`.
+	//
+	// If in doubt bump it. This will lead to removal of the existing artifacts in the on-disk cache
+	// and recompilation.
+}
+
+pub enum ArtifactState {
+	/// The artifact is ready to be used by the executor.
+	///
+	/// That means that the artifact should be accessible through the path obtained by the artifact
+	/// id under (unless, it was removed externally).
+	Prepared {
+		/// The time when the artifact was the last time needed.
+		///
+		/// This is updated when we get the heads up for this artifact or when we just discover
+		/// this file.
+		last_time_needed: SystemTime,
+	},
+	/// A task to prepare this artifact is scheduled.
+	Preparing,
+}
+
+pub struct Artifacts {
+	artifacts: HashMap<ArtifactId, ArtifactState>,
+}
+
+impl Artifacts {
+	/// Scan the given cache root for the artifacts.
+	///
+	/// The recognized artifacts will be filled in the table and unrecognized will be removed.
+	pub async fn new(cache_path: &Path) -> Self {
+		// Make sure that the cache path directory and all it's parents are created.
+		let _ = async_std::fs::create_dir_all(cache_path).await;
+
+		let artifacts = match scan_for_known_artifacts(cache_path).await {
+			Ok(a) => a,
+			Err(err) => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					"unable to seed the artifacts in memory cache: {:?}. Starting with a clean one",
+					err,
+				);
+				HashMap::new()
+			}
+		};
+
+		Self { artifacts }
+	}
+
+	#[cfg(test)]
+	pub(crate) fn empty() -> Self {
+		Self {
+			artifacts: HashMap::new(),
+		}
+	}
+
+	/// Returns the state of the given artifact by its ID.
+	pub fn artifact_state_mut(&mut self, artifact_id: &ArtifactId) -> Option<&mut ArtifactState> {
+		self.artifacts.get_mut(artifact_id)
+	}
+
+	/// Inform the table about the artifact with the given ID. The state will be set to "preparing".
+	pub fn insert_preparing(&mut self, artifact_id: ArtifactId) {
+		always!(self
+			.artifacts
+			.insert(artifact_id, ArtifactState::Preparing)
+			.is_none());
+	}
+
+	/// Insert an artifact with the given ID as "prepared".
+	#[cfg(test)]
+	pub fn insert_prepared(&mut self, artifact_id: ArtifactId, last_time_needed: SystemTime) {
+		always!(self
+			.artifacts
+			.insert(artifact_id, ArtifactState::Prepared { last_time_needed })
+			.is_none());
+	}
+
+	/// Remove and retrive the artifacts from the table that are older than the supplied Time-To-Live.
```

Comment:
We don't remove in here?

Reply:
Ugh, the comment is right and removal was lost during a refactoring. Fixed it and modified the test.